# Cleaning data - sharks dataframe

# Paso 1: importar dataframe y hacer una primera observación de los datos que la componen

In [1701]:
import pandas as pd
import os
from re import sub
import numpy
from statistics import mean
%matplotlib inline
import matplotlib.pyplot as plt

En primer lugar, importo el dataframe con el que voy a trabajar para poder empezar a trabajar con los datos.

In [1702]:
data = pd.read_csv('attacks.csv', encoding='latin-1')
data.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09         6301.0         NaN         NaN  
3    2018.06.08         6300.0         NaN         NaN  
4    2018.06.04         6299.0         NaN         NaN  

[5 rows x 24 columns]

Después de haber importado este dataframe, busco resolver algunas primeras dudas (¿Qué columnas la componen?, ¿Cuántos datos continenen?, ¿Cuántos nulos hay por columna?, etc.)

In [1703]:
#Número de filas y columnas
data.shape

(25723, 24)

In [1704]:
#Nombre de las columnas
data.columns

/home/marta/.local/lib/python3.6/site-packages/IPython/core/displayhook.py:271: UserWarning: Output cache limit (currently 1000 entries) hit.
Flushing oldest 200 entries.
  'Flushing oldest {cull_count} entries.'.format(sz=sz, cull_count=cull_count))


Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [1705]:
#tipo de datos
data.dtypes.head()

Case Number     object
Date            object
Year           float64
Type            object
Country         object
dtype: object

In [1706]:
#También me interesa saber cuantos nulos tengo en cada variable
data.isnull().sum().head()

Case Number    17021
Date           19421
Year           19423
Type           19425
Country        19471
dtype: int64

Creo una copia de la base de datos original, para trabajar desde esa copia (tiburon):

In [1707]:
tiburon = data
tiburon.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09         6301.0         NaN         NaN  
3    2018.06.08         6300.0         NaN         NaN  
4    2018.06.04         6299.0         NaN         NaN  

[5 rows x 24 columns]

# Hipótesis


Después de haber observado el DataFrame y el tipo de información que la integra, considero que sería interesante realizar mi trabajo con respecto a la siguiente hipótesis: Desde 1950, el número de ataques de tiburón tienen consecuencias menos importantes para sus víctimas con el paso de los años.

# Paso 2: Hacer la limpieza de datos

Aunque existen muchas maneras de limpiar datos, he decidido quedarme exclusivamente con la información que me permite realizar el trabajo de forma directa. He tomado esta decisión ya que considero que la recogida de datos contiene muchos errores y tiene poca fiabilidad. Por esta razón, me parece peligroso basarme en otras columnas que no sean aquellas que necesito directamente para mi trabajo. Veremos hasta qué punto los resultados que voy a obtener son significativos, con respecto a la muestra con la que voy a operar.

In [1708]:
#Elimino las columnas con las que no voy a operar, creando la variable col_malas

col_malas = tiburon.drop(columns = ['Date','Type','Sex ','Country','Area','Location','Activity','Name','Age','Injury','Time','Species ','Investigator or Source','pdf','href formula','href','Case Number.1','Case Number.2','original order','Unnamed: 22','Unnamed: 23'], axis=1)
tiburon = col_malas
tiburon.head()



Case Number    Year Fatal (Y/N)
0  2018.06.25  2018.0           N
1  2018.06.18  2018.0           N
2  2018.06.09  2018.0           N
3  2018.06.08  2018.0           N
4  2018.06.04  2018.0           N

No es necesario cambiar el nombre de nuestras columnas ya que no tienen errores (contrariamente al nombre de otras columnas que hemos borrado).

Para entrenarme a crear funciones, y sabiendo que voy a tener que limpiar muchos datos, he decidido crear una función para conocer número de valores de todas las columnas:

In [1709]:
def cls_counts(df):
    for col in df.columns:
        print("---%s---" %col)
        print(df[col].value_counts())

In [1710]:
#Pruebo mi función:
cls_counts(tiburon)

---Case Number---
0                 2400
2009.12.18           2
2013.10.05           2
1907.10.16.R         2
1980.07.00           2
1913.08.27.R         2
1920.00.00.b         2
2014.08.02           2
2005.04.06           2
1923.00.00.a         2
1990.05.10           2
2006.09.02           2
1966.12.26           2
1915.07.06.a.R       2
2012.09.02.b         2
1962.06.11.b         2
1983.06.15           2
1946.00.00.a         1
2017.08.29           1
1965.12.19           1
1878.09.02.a         1
1950.04.09.b         1
1987.06.13           1
1960.08.10           1
2016.04.09           1
1965.12.10           1
1996.01.12           1
2000.06.09.a         1
2014.08.29.b         1
1958.12.13           1
                  ... 
1919.03.16           1
1994.00.00.a         1
2018.03.09.a         1
1973.11.25           1
1939.10.04           1
1963.04.22           1
1948.11.18           1
1936.04.22           1
2001.08.25           1
1901.00.00           1
1928.05.17           1
1895.00.00.b    

Ahora voy a centrarme en limpiar por separado aquellas columnas que componen mi nueva tabla.

# A) Columna Case Number

En primer lugar, quiero eliminar las filas duplicadas en la columna Case Number. Esta columna es importante ya que nos da una estimación del número total de ataques de tiburón.

Después de haber observado el dataframe, he llegado a la conclusión de que cada una de las filas debería tener un valor diferente: vemos que cada registro está compuesto de la fecha del día del accidente (AAAA.MM.DD), a la que se le añade una letra en orden alfabético en los casos en los que ha habido más de un ataque en el día. 

Por supuesto, no tenemos la certeza de que los registros se nombren de esta manera. Aún mucho menos que quitando los duplicados, obtengamos el número real de ataques de tiburón. No obstante, teniendo tan poca información acerca de cómo se han recogido y plasmado los datos, ninguna metodología nos daría esa seguridad.

In [1711]:
#Eliminamos los registros duplicados si los hubiese
var1 = tiburon.drop_duplicates(subset=['Case Number'], keep='first')
tiburon['Case Number'] = var1
tiburon.shape


(25723, 3)

Vemos que no existían duplicados, ya que el número de filas se ha mantenido. Ahora vamos a proceder a cambiar el valor de los registros nulos por 0, ya que vamos a proceder a eliminarlos. 

In [1712]:
var2 = tiburon.dropna(how = 'any')
tiburon = var2
tiburon.shape

(5749, 3)

Ahora el dataframe con el que vamos a trabajar va a ser mucho más reducido, ya que hemos descendido de 25723 a 5749 filas.

## B) Columna Year

Elimino todos aquellos registros en los que su valor sea cero o nulo. También modifico el tipo de Year a un 'interger', para eliminar el '.0'.

In [1713]:
tiburon["Year"].isnull()
pd.to_numeric(tiburon["Year"], errors = "coerce")
tiburon = tiburon.dropna(subset = ["Year"])
tiburon["Year"] = tiburon["Year"].astype(int)

Ahora vamos a eliminar las filas que incluyan fechas inferiores a 1950.

In [1714]:
var3 = tiburon[(tiburon["Year"] >= 1950)]
tiburon = var3
tiburon.head()

Case Number  Year Fatal (Y/N)
0  2018.06.25  2018           N
1  2018.06.18  2018           N
2  2018.06.09  2018           N
3  2018.06.08  2018           N
4  2018.06.04  2018           N

## C) Columna Fatal (Y/N)

Además de lo hecho anteriormente, vamos a eliminar las filas en las que el valor en las columnas Year y Fatal (Y/N) sea igual a cero. Esto lo hacemos para que estos datos (que parecen incorrectos), no afecten a nuestros resultados.

In [1715]:
tiburon = tiburon[tiburon["Fatal (Y/N)"] != 0].copy()
tiburon.shape


(4123, 3)

Nuestra muestra ha descendido mucho, tal vez demasiado para que podamos sacar conclusiones significativas.

In [1716]:
tiburon.head()

Case Number  Year Fatal (Y/N)
0  2018.06.25  2018           N
1  2018.06.18  2018           N
2  2018.06.09  2018           N
3  2018.06.08  2018           N
4  2018.06.04  2018           N

Ahora vamos a homogeneizar los registros de manera a dejar tres opciones posibles (Y, N, Unknown). Como he encontrado muchos problemas al intentar resolver esta cuestión con replace y con regex, al final he optado por hacer un diccionario.

In [1717]:
def homog_fatal (v):

    fatal = {

        "Y":"Y", "N":"N", "UNKNOWN":"UNKNOWN","y":"Y", " Y":"Y", "Y ":"Y", "n":"N", "N ":"N", " N":"N","M":"UNKNOWN","2017":"UNKNOWN"

    }

    for fatal, fila in fatal.items(): 

        if fatal in v: 

            return fila

    return None

In [1718]:
tiburon["Fatal (Y/N)"] = tiburon["Fatal (Y/N)"].apply(homog_fatal)   
tiburon["Fatal (Y/N)"].value_counts()

N          3491
Y           630
UNKNOWN       2
Name: Fatal (Y/N), dtype: int64

No obstante, como oodemos observar aquí, nos encontramos ante un error que no logro resolver. Solamente aparecen 2 filas con el nombre "UNKNOWN", a pesar de que anteriormente contábamos con 72 filas con ese valor, y que en el diccionario hemos incluído la posibilidad de que las filas tengan ese nombre.

In [1719]:
display(tiburon.head())

Case Number  Year Fatal (Y/N)
0  2018.06.25  2018           N
1  2018.06.18  2018           N
2  2018.06.09  2018           N
3  2018.06.08  2018           N
4  2018.06.04  2018           N

# Paso 3: Realizar operaciones estadísticas para averiguar si la hipótesis que hemos planteado se cumple.

In [1720]:
cls_counts(tiburon)

---Case Number---
1983.07.05      1
1955.08.26      1
1969.05.25      1
2017.11.24      1
1981.12.13.a    1
1960.01.21      1
2014.07.14      1
1959.01.12      1
1966.08.16      1
2013.06.14.R    1
1995.06.30      1
2016.09.18.a    1
2004.09.04      1
1966.07.17.b    1
1965.04.04.b    1
2010.07.17.a    1
1992.05.00      1
2015.05.29.b    1
2018.05.27      1
2013.07.02      1
1960.04.00.b    1
2003.10.05.c    1
1970.12.07      1
1998.08.23.R    1
2018.04.23      1
1987.05.08      1
2003.04.15      1
2001.09.13      1
2007.01.25.b    1
2012.09.16      1
               ..
2014.12.29.b    1
2015.02.15      1
1991.11.19      1
2017.06.10.b    1
1986.09.00.a    1
2012.09.24      1
2014.10.17      1
2006.09.02      1
1987.02.19      1
2017.07.29.a    1
2007.03.31.a    1
1994.08.22      1
1991.07.01.b    1
1955.04.13.R    1
2016.06.27      1
1997.01.01      1
1977.04.21      1
2007.08.09.b    1
2006.10.31      1
1966.12.27.b    1
2010.07.23      1
1953.07.31      1
1975.07.05      1
1977.08.23

In [1721]:
tiburon.dtypes


Case Number    object
Year            int64
Fatal (Y/N)    object
dtype: object

A través de la tabla que presentamos a continuación, podemos observar el número total de ataques fatales "Y", ataques fatales "N" y los ataques de los que no se tienen datos, por año, desde 1950.

In [1722]:
print(pd.crosstab(tiburon["Year"],tiburon["Fatal (Y/N)"]).head())

Fatal (Y/N)   N  UNKNOWN   Y
Year                        
1950         29        0  10
1951         18        0  11
1952         18        0   9
1953         29        0   6
1954         23        0  15


Ahora vamos a hacer la misma operación pero en porcentajes.

In [1723]:
print(pd.crosstab(tiburon["Year"],tiburon["Fatal (Y/N)"],normalize = "index"))

Fatal (Y/N)         N   UNKNOWN         Y
Year                                     
1950         0.743590  0.000000  0.256410
1951         0.620690  0.000000  0.379310
1952         0.666667  0.000000  0.333333
1953         0.828571  0.000000  0.171429
1954         0.605263  0.000000  0.394737
1955         0.714286  0.000000  0.285714
1956         0.588235  0.000000  0.411765
1957         0.736842  0.000000  0.263158
1958         0.625000  0.000000  0.375000
1959         0.703704  0.000000  0.296296
1960         0.825000  0.000000  0.175000
1961         0.819444  0.000000  0.180556
1962         0.810127  0.000000  0.189873
1963         0.745455  0.000000  0.254545
1964         0.724138  0.000000  0.275862
1965         0.790698  0.000000  0.209302
1966         0.700000  0.000000  0.300000
1967         0.756098  0.000000  0.243902
1968         0.864865  0.000000  0.135135
1969         1.000000  0.000000  0.000000
1970         0.800000  0.000000  0.200000
1971         0.714286  0.000000  0

In [1724]:
#No aparece Values como Case Number... ¿Por qué?
print(tiburon.pivot_table(index= ["Year"], columns= ["Fatal (Y/N)"],values=["Case Number"], aggfunc = pd.Series.mean))

               Year                
Fatal (Y/N)       N UNKNOWN       Y
Year                               
1950         1950.0     NaN  1950.0
1951         1951.0     NaN  1951.0
1952         1952.0     NaN  1952.0
1953         1953.0     NaN  1953.0
1954         1954.0     NaN  1954.0
1955         1955.0     NaN  1955.0
1956         1956.0     NaN  1956.0
1957         1957.0     NaN  1957.0
1958         1958.0     NaN  1958.0
1959         1959.0     NaN  1959.0
1960         1960.0     NaN  1960.0
1961         1961.0     NaN  1961.0
1962         1962.0     NaN  1962.0
1963         1963.0     NaN  1963.0
1964         1964.0     NaN  1964.0
1965         1965.0     NaN  1965.0
1966         1966.0     NaN  1966.0
1967         1967.0     NaN  1967.0
1968         1968.0     NaN  1968.0
1969         1969.0     NaN     NaN
1970         1970.0     NaN  1970.0
1971         1971.0     NaN  1971.0
1972         1972.0     NaN  1972.0
1973         1973.0     NaN  1973.0
1974         1974.0     NaN 

In [1725]:
tiburon.hist(column="Fatal (Y/N)", by "Year")

SyntaxError: invalid syntax (<ipython-input-1725-6d708067a222>, line 1)

In [ ]:
#Por qué no sale nada?
tiburon.boxplot(column="Fatal (Y/N)", by "Year")

In [ ]:
#Por qué sale solo year?
stats = tiburon.describe().transpose()
tiburon = stats['IQR'] = stats['75%'] - stats['25%']
tiburon

In [ ]:
pd.tools.plotting.scatter_matrix(tiburon.select_dtypes(exclude = ["object"]))